### PR TITLE
Add Lyric support

### PR DESCRIPTION
**Pull request by robert-hamilton36**
Changes
Adds route for lyrics. Will show current lyric highlighted if saved file allows it
Add item detail page for individual songs that show the lyrics

Previews

#/Lyrics
image

#/details?id={songsitemId}
image

### CODE REVIEW

**Comment by ebkalderon** [comment]
Hey there! Just wondering what other work might need to be done before this PR can be taken out of draft and reviewed. The implementation itself looks great. Might this be blocked on something that needs additional help shepherding through?

**Comment by robert-hamilton36** [comment]
Oh no, Sorry. I didn't mean to leave it in draft. I think its good to go and be reviewed.

**Comment by thornbill** [comment]
I believe this will require changes to be compatible with https://github.com/jellyfin/jellyfin/pull/9951

File: src/controllers/itemDetails/index.js
Comment (dmitrylyzo) [inline src/controllers/itemDetails/index.js:44]: I believe we can now use the absolute import path.
```suggestion
import 'styles/lyrics.scss';
import 'styles/scrollstyles.scss';
```

File: src/controllers/itemDetails/index.js
Comment (dmitrylyzo) [inline src/controllers/itemDetails/index.js:1096]: Escape HTML `lyric.Text`?
Unless the lyric has HTML styling, but that's dangerous.

File: src/controllers/itemDetails/index.js
Comment (dmitrylyzo) [inline src/controllers/itemDetails/index.js:1309]: I would leave this empty line for readability.

File: src/controllers/lyrics.js
Comment (dmitrylyzo) [inline src/controllers/lyrics.js:18]: Again, shouldn't we escape HTML `lyric.Text`? Here and below.

File: src/controllers/playback/queue/index.html
Comment (dmitrylyzo) [inline src/controllers/playback/queue/index.html:88]:
```suggestion
                        <button is="paper-icon-button-light" class="btnShuffleQueue autoSize" title="${Shuffle}">
```

File: src/controllers/playback/queue/index.html
Comment (dmitrylyzo) [inline src/controllers/playback/queue/index.html:84]:
```suggestion
                        <button is="paper-icon-button-light" class="btnLyrics autoSize hide" title="${Lyrics}">
```

File: src/scripts/deleteHelper.js
Comment (dmitrylyzo) [inline src/scripts/deleteHelper.js:51]:
```suggestion
    deleteItem,
    deleteLyrics
```

**Review by dmitrylyzo — COMMENTED** [review]
I reviewed without testing.

File: src/components/itemContextMenu.js
Comment (robert-hamilton36) [reply src/components/itemContextMenu.js]: Thanks, apologies

File: src/components/itemContextMenu.js
Comment (robert-hamilton36) [reply src/components/itemContextMenu.js:655]: Yeah this works. Should I also update line 640  deleteItems (which this was based on) to the same format?

File: src/components/nowPlayingBar/nowPlayingBar.js
Comment (robert-hamilton36) [reply src/components/nowPlayingBar/nowPlayingBar.js:88]: Sweet. I will do the whole file as well on a different branch. I'll add the title for this button on this branch however

File: src/components/itemContextMenu.js
Comment (robert-hamilton36) [reply src/components/itemContextMenu.js:621]: In most cases I believe this caused two navigations.

Navigation was also being handled by the methods that called itemContextMenu.show().

Although I think this change might be better off in a different branch, I'm looking at. So I will revert this.

On that topic, scripts/deleteHelper.js is only used in itemContextMenu.js. Is this something we should consider moving entirely into itemContextMenu instead of having a separate script

File: src/controllers/itemDetails/index.html
Comment (robert-hamilton36) [reply src/controllers/itemDetails/index.html:192]: Ok right, I think focuscontainer-x must have been a remnant from which ever section I decided to copy for a skeleton. I never managed to work out what it did, So kept it in in case it was important. I will remove it.

File: src/controllers/lyrics.js
Comment (robert-hamilton36) [reply src/controllers/lyrics.js]: fixed on [a1768dd](https://github.com/jellyfin/jellyfin-web/pull/4733/commits/a1768dd4df99aaa428d8ef4ebc03c4d288e837fc)

File: src/controllers/itemDetails/index.js
Comment (robert-hamilton36) [reply src/controllers/itemDetails/index.js]: fixed on [a1768dd](https://github.com/jellyfin/jellyfin-web/pull/4733/commits/a1768dd4df99aaa428d8ef4ebc03c4d288e837fc)

File: src/controllers/itemDetails/index.js
Comment (robert-hamilton36) [reply src/controllers/itemDetails/index.js]: fixed on [f84ca77](https://github.com/jellyfin/jellyfin-web/pull/4733/commits/f84ca7711f789cb4065adef576e90f7117436792)

File: src/controllers/itemDetails/index.js
Comment (robert-hamilton36) [reply src/controllers/itemDetails/index.js]: Thanks! Styles aren't currently being used in itemDetails so this import has been removed (I forgot to remove it when I changed it)

Fixed [17c61f4](https://github.com/jellyfin/jellyfin-web/pull/4733/commits/17c61f4b45928a188276c88a0cc05b9d3a94e23a)

File: src/controllers/itemDetails/index.html
Comment (robert-hamilton36) [reply src/controllers/itemDetails/index.html]: Fixed on [063f16a](https://github.com/jellyfin/jellyfin-web/pull/4733/commits/063f16aaae784a19bb4dd7606ae066da89818df3)

File: src/components/nowPlayingBar/nowPlayingBar.js
Comment (robert-hamilton36) [reply src/components/nowPlayingBar/nowPlayingBar.js]: single title for lyrics added here [9e51b5c](https://github.com/jellyfin/jellyfin-web/pull/4733/commits/9e51b5c0ddead78a2b1bc41133b5417b45b37b8b)

**Comment by thornbill** [comment]
You also need to rebase or merge master back into this branch to fix the conflicts. We are still interested in getting this in for 10.9.0 if we can get it reviewed and any issues resolved this week.

File: src/controllers/itemDetails/index.js
Comment (dmitrylyzo) [inline src/controllers/itemDetails/index.js:1163]: AFAIK, this field has been removed.
```suggestion
            Fields: 'ItemCounts,PrimaryImageAspectRatio,CanDelete,MediaSourceCount',
```

File: src/controllers/lyrics.html
Comment (dmitrylyzo) [inline src/controllers/lyrics.html:6]: Remove trailing spaces.

**Review by dmitrylyzo — COMMENTED** [review]
Another pass.

1. Shouldn't we close the lyrics page after music stops?

2. This is not good in TV display mode as there is no scrolling. In fact, we don't handle text blocks in TV display mode well.

3. In Mobile display mode, the `Lyrics` button is small. This may be improved in the future. Maybe after a redesign or so.
lyrics-small-button

**Comment by robert-hamilton36** [comment]
> Another pass.
> 
>     1. Shouldn't we close the lyrics page after music stops?
> 
>     2. This is not good in TV display mode as there is no scrolling. In fact, we don't handle text blocks in TV display mode well.
> 
>     3. In Mobile display mode, the `Lyrics` button is small. This may be improved in the future. Maybe after a redesign or so.
>        lyrics-small-button
> 
> 
> UPD: 4. There is a new section called `More From This Album`. It doesn't seem to be related to the lyrics feature. Maybe it should be moved to the new PR?

1. Yes that would be a good idea. I'm thinking i'll just have to add an event for 'playbackstop' that appRouter.back()
2. Right silly me I guess I must have always used my scroll wheel, not realizing tvs couldn't do that. Assuming the keyboards arrow keys act the same as the tvs remotes navigation, I'll look into add scrolling. I'm not familiar with how its been implemented already in menus
3. Hmm yes, secondary buttons are hidden in mobile mode. I didn't realize and left the lyric button visible. I'll hide it, the problem is getting back onto that lyric page. For now, what I can do is add that lyric button to the item details page like so


Add Mic

Which would require accessing view lyrics from the context menu and than clicking the mic button

4. Yes its not directly related to the lyric feature. My thinking was, that the new section is only relevant because of the new route #/details?id={songsitemId} that is opened on the view lyrics context menu. But will move into a seperate PR

**Comment by robert-hamilton36** [comment]
> Another pass.
> 
>     1. Shouldn't we close the lyrics page after music stops?
> 
>     2. This is not good in TV display mode as there is no scrolling. In fact, we don't handle text blocks in TV display mode well.
> 
>     3. In Mobile display mode, the `Lyrics` button is small. This may be improved in the future. Maybe after a redesign or so.
>        lyrics-small-button
> 
> 
> UPD: 4. There is a new section called `More From This Album`. It doesn't seem to be related to the lyrics feature. Maybe it should be moved to the new PR?

1. fixed with https://github.com/jellyfin/jellyfin-web/pull/4733/commits/73c01fb688481c429f6d42092ec3ce3f2b6ff7f3
2. possible fix suggested above 
3. fixed with https://github.com/jellyfin/jellyfin-web/pull/4733/commits/cd8d5c73de2d10c9b350dadcbc67f4cfe75983f7 and then https://github.com/jellyfin/jellyfin-web/pull/4733/commits/83bee29bd7c375a8804b71b34091180ff49453b6 
4. fixed with https://github.com/jellyfin/jellyfin-web/pull/4733/commits/286e0bf4f8e82897d7d87f3ad5b80304bd50b464

**Comment by robert-hamilton36** [comment]
> You also need to rebase or merge master back into this branch to fix the conflicts. We are still interested in getting this in for 10.9.0 if we can get it reviewed and any issues resolved this week.

I hope this was it https://github.com/jellyfin/jellyfin-web/pull/4733/commits/f1d955a941ce069b322cad69525a1b4f497d786f

Or else I might need a bit of guidance

File: src/components/playback/playbackmanager.js
Comment (dmitrylyzo) [inline src/components/playback/playbackmanager.js:1643]: Ideally, we should receive this event after the seeking is complete.
I believe that `timeupdate` is the correct event.

**Comment by robert-hamilton36** [comment]
I have found an issue. `realeaseCurrentPlayer()` doesn't seem to be unbinding events. So even after you exit the lyrics page the events are still being run in the background

File: src/controllers/lyrics.js
Comment (dmitrylyzo) [inline src/controllers/lyrics.js:127]:
```suggestion
    function getLyrics(serverId, itemId) {
```

File: src/controllers/lyrics.js
Comment (dmitrylyzo) [inline src/controllers/lyrics.js:142]:
```suggestion
    function bindToPlayer(player) {
```

File: src/controllers/lyrics.js
Comment (dmitrylyzo) [inline src/controllers/lyrics.js:181]: What's the problem with `getCurrentPlayTime()`?
I don't think we should modify the `timeupdate` event for no reason. _In fact, it has data that differs from player to player - not good, imo._

**Review by dmitrylyzo — COMMENTED** [review]
The only thing that bothers me is this behavior:
1. Start music.
2. Click the now playing bar (open queue).
3. Open lyrics.
4. Stop music using the now playing bar.
5. The lyrics page executes `appRouter.back`.
6. The queue page appears to be broken.

**Comment by robert-hamilton36** [comment]
> The only thing that bothers me is this behavior:
> 
>     1. Start music.
> 
>     2. Click the now playing bar (open queue).
> 
>     3. Open lyrics.
> 
>     4. Stop music using the now playing bar.
> 
>     5. The lyrics page executes `appRouter.back`.
> 
>     6. The queue page appears to be broken.

Yes so I think appRouter.goHome() is the best option. Because if you stay on the lyric page, the arrow in the top left will still take you to the broken queue page

File: src/controllers/lyrics.js
Comment (dmitrylyzo) [inline src/controllers/lyrics.js:193]: The lyrics page will be kept until there is a next item and will be closed at the end of the playing queue.
```suggestion
    function onPlaybackStop (_, state) {
        if (!state.NextMediaType) {
            appRouter.back();
        }
    }
```

**Comment by dmitrylyzo** [comment]
> Yes so I think appRouter.goHome() is the best option. Because if you stay on the lyric page, the arrow in the top left will still take you to the broken queue page

This workaround seems to work:
```diff
diff --git a/src/controllers/playback/queue/index.js b/src/controllers/playback/queue/index.js
index d95081b91..60c541bea 100644
--- a/src/controllers/playback/queue/index.js
+++ b/src/controllers/playback/queue/index.js
@@ -1,4 +1,5 @@
 import RemoteControl from '../../../components/remotecontrol/remotecontrol';
+import { appRouter, history } from 'components/router/appRouter';
 import { playbackManager } from '../../../components/playback/playbackmanager';
 import libraryMenu from '../../../scripts/libraryMenu';
 import '../../../elements/emby-button/emby-button';
@@ -35,8 +36,19 @@ export default function (view) {
     }
 
     view.addEventListener('viewshow', function () {
+        const player = playbackManager.getCurrentPlayer();
+
+        if (!player) {
+            if (history.length > 1) {
+                appRouter.back();
+            } else {
+                appRouter.goHome();
+            }
+            return;
+        }
+
         libraryMenu.setTransparentMenu(true);
-        bindToPlayer(playbackManager.getCurrentPlayer());
+        bindToPlayer(player);
         document.addEventListener('keydown', onKeyDown);
 
         if (remoteControl) {
```
I hope it doesn't break anything.

**Comment by robert-hamilton36** [comment]
> > Yes so I think appRouter.goHome() is the best option. Because if you stay on the lyric page, the arrow in the top left will still take you to the broken queue page
> 
> This workaround seems to work:
> 
> ```diff
> diff --git a/src/controllers/playback/queue/index.js b/src/controllers/playback/queue/index.js
> index d95081b91..60c541bea 100644
> --- a/src/controllers/playback/queue/index.js
> +++ b/src/controllers/playback/queue/index.js
> @@ -1,4 +1,5 @@
>  import RemoteControl from '../../../components/remotecontrol/remotecontrol';
> +import { appRouter, history } from 'components/router/appRouter';
>  import { playbackManager } from '../../../components/playback/playbackmanager';
>  import libraryMenu from '../../../scripts/libraryMenu';
>  import '../../../elements/emby-button/emby-button';
> @@ -35,8 +36,19 @@ export default function (view) {
>      }
>  
>      view.addEventListener('viewshow', function () {
> +        const player = playbackManager.getCurrentPlayer();
> +
> +        if (!player) {
> +            if (history.length > 1) {
> +                appRouter.back();
> +            } else {
> +                appRouter.goHome();
> +            }
> +            return;
> +        }
> +
>          libraryMenu.setTransparentMenu(true);
> -        bindToPlayer(playbackManager.getCurrentPlayer());
> +        bindToPlayer(player);
>          document.addEventListener('keydown', onKeyDown);
>  
>          if (remoteControl) {
> ```
> 
> I hope it doesn't break anything.


If you manually enter `/#/queue` into the browser.
The queue still has the playbar but with 3 added menus:
1. Navigation
2. Send message
3. Enter text

They don't seem to do anything, but I don't know if its still in use or not.
But this fix will stop you from getting to this menu.


Queue

**Comment by dmitrylyzo** [comment]
> If you manually enter `/#/queue` into the browser.
> The queue still has the playbar but with 3 added menus:

When I tested, it redirected back if there was something open, or went to the home page when entered in a blank tab.

> But this fix will stop you from getting to this menu.

I haven't tested remote control. This page should be accessible from the `PlayOn` button -> Remote control.

**Comment by robert-hamilton36** [comment]
I've reverted `onPlaybackStop` Behaviour to `goHome()`, until a work around is found

File: src/controllers/lyrics.js
Comment (dmitrylyzo) [inline src/controllers/lyrics.js:201]: Have you tried playing a playlist? I believe when the song reaches the end, the player emits a stop event. This way, you will be taken to the home page each time.

You can prevent this by checking `state.NextMediaType` (https://github.com/jellyfin/jellyfin-web/pull/4733#discussion_r1542632714).

File: src/components/itemContextMenu.js
Comment (nielsvanvelzen) [inline src/components/itemContextMenu.js:534]:
```suggestion
                    appRouter.show('lyrics');
```

File: src/components/nowPlayingBar/nowPlayingBar.js
Comment (nielsvanvelzen) [inline src/components/nowPlayingBar/nowPlayingBar.js:226]:
```suggestion
            appRouter.show('lyrics');
```

File: src/components/nowPlayingBar/nowPlayingBar.js
Comment (nielsvanvelzen) [inline src/components/nowPlayingBar/nowPlayingBar.js:793]: Not the best way to check because the paths are case insensitive.

File: src/components/remotecontrol/remotecontrol.js
Comment (nielsvanvelzen) [inline src/components/remotecontrol/remotecontrol.js:775]:
```suggestion
            appRouter.show('lyrics');
```

File: src/components/nowPlayingBar/nowPlayingBar.js
Comment (dmitrylyzo) [reply src/components/nowPlayingBar/nowPlayingBar.js:793]: Quick fix. Lyrics toggling doesn't work.
```suggestion
    lyricPageActive = appRouter.currentRouteInfo.path.toLowerCase() === '/lyrics';
```

File: src/components/nowPlayingBar/nowPlayingBar.js
Comment (dmitrylyzo) [inline src/components/nowPlayingBar/nowPlayingBar.js:826]: Quick fix. Lyrics toggling doesn't work.
```suggestion
    lyricPageActive = appRouter.currentRouteInfo.path.toLowerCase() === '/lyrics';
```

File: src/styles/lyrics.scss
Comment (dmitrylyzo) [inline src/styles/lyrics.scss:10]: This doesn't work on webOS 1.2 (old WebKit). You can reproduce the looking by removing `height: max-content;`.
The only way I found is to set `height` for the lyrics element.

**Review by dmitrylyzo — COMMENTED** [review]
_Oops, I forgot to submit review._

File: src/controllers/lyrics.html
Comment (dmitrylyzo) [inline src/controllers/lyrics.html:3]: For some reason `flex` inside another `flex` has the wrong height - the now playing bar overlaps the last line.
Placing it in another `div` fixes the problem.
```suggestion
    <div>
        <div class="dynamicLyricsContainer padded-bottom-page">
        </div>
    </div>
```

File: src/controllers/lyrics.js
Comment (dmitrylyzo) [inline src/controllers/lyrics.js:200]:
```suggestion
    function onPlaybackStop(_, state) {
```

File: src/controllers/lyrics.js
Comment (dmitrylyzo) [inline src/controllers/lyrics.js:237]:
```suggestion
    view.addEventListener('viewshow', function () {
```

File: src/controllers/lyrics.js
Comment (dmitrylyzo) [inline src/controllers/lyrics.js:246]:
```suggestion
    view.addEventListener('viewbeforehide', function () {
```

**Review by dmitrylyzo — COMMENTED** [review]
I think these are the last ones. :crossed_fingers: 

This PR requires squashing before merging.

File: src/components/nowPlayingBar/nowPlayingBar.js
Comment (thornbill) [inline src/components/nowPlayingBar/nowPlayingBar.js:89]: Why not use the lyrics icon?

File: src/controllers/playback/queue/index.html
Comment (thornbill) [inline src/controllers/playback/queue/index.html:85]: Same here.

File: src/components/nowPlayingBar/nowPlayingBar.js
Comment (dmitrylyzo) [reply src/components/nowPlayingBar/nowPlayingBar.js:89]: `lyrics` icon conflicts with new `lyrics` class. But it looks better.

**Review by dmitrylyzo — APPROVED** [review]
LGTM.

Who will squash the commits?

**Comment by robert-hamilton36** [comment]
I can squash, assuming this is alright
`git reset --soft master`
`git push-f master LyricsSupport`

**Comment by dmitrylyzo** [comment]
> I can squash, assuming this is alright
> `git reset --soft master`
> `git push-f master LyricsSupport`

I don't think this will collect all the changes because merge commits (instead of rebase) were used.

There may be an easier way, but I am used to do it this way (being on the PR branch):
1. Update the upstream.
`git fetch upstream`, where `upstream` is a remote pointing to the upstream repo of jellyfin-web.
2. Backup current branch. To compare with it later.
`git branch backup-rebase`
3. Run interactive rebase.
`git rebase -i upstream/master`.
In the rebase script, you must `pick` the first commit, others - `fixup`.
4. Fix conflicts if any. Then, `git rebase --continue`. Until rebase is complete.
You can abort rebase with `git rebase --abort`.
6. Compare the current branch with the backup to see if there are any unnecessary changes.
7. If it is fine, force push.
If you ruin your branch, you can re-checkout it from the backup.

**Comment by robert-hamilton36** [comment]
> > I can squash, assuming this is alright
> > `git reset --soft master`
> > `git push-f master LyricsSupport`
> 
> I don't think this will collect all the changes because merge commits (instead of rebase) were used.
> 
> There may be an easier way, but I am used to do it this way (being on the PR branch):
> 
>     1. Update the upstream.
>        `git fetch upstream`, where `upstream` is a remote pointing to the upstream repo of jellyfin-web.
> 
>     2. Backup current branch. To compare with it later.
>        `git branch backup-rebase`
> 
>     3. Run interactive rebase.
>        `git rebase -i upstream/master`.
>        In the rebase script, you must `pick` the first commit, others - `fixup`.
> 
>     4. Fix conflicts if any. Then, `git rebase --continue`. Until rebase is complete.
>        You can abort rebase with `git rebase --abort`.
> 
>     5. Compare the current branch with the backup to see if there are any unnecessary changes.
> 
>     6. If it is fine, force push.
>        If you ruin your branch, you can re-checkout it from the backup.

Awesome! Thanks so much for all your help

**Comment by ebkalderon** [comment]
Very excited to see this feature finally land in the 10.9 release, and not a moment too soon! Thanks to all involved. :tada:

**Review by thornbill — APPROVED** [review]

**Pull request merged**
